### ISSUE TITLE
 SCSS @use vs CSS @layer Problems.... plz fixed.. this issues..

### ISSUE THREAD

**Issue by rebehayan** [sass/sass#3842]
While using CSS's @layer in scss, 
I found that it did not work in files that had to be compiled at the top level.

The contents of my compiled file are as follows.

//style.scss
@layer reset, components, utility;
@use 'components';
@use 'pages';

The problem in that code is
@use must be at the top of the code
This problem seems to have arisen because it is similar to the @layer concept in CSS.

Is there a way to fix this problem or can it be fixed in a future update?

---

I wrote this issue through a translator.
Some expressions may be strange.

**Comment by nex3** [comment]
We generally recommend wrapping partials in mixins, so that instead of having your `@use` rules emit styles as a side-effect you can choose exactly where and when they're emitted. For example:

```scss
@use 'components';
@use 'pages';

@layer reset, components, utility;

@include components.styles;
@include pages.styles;
```

That said, we should probably move the `@layer` rule to the top of the output stylesheet as we do for plain-CSS `@import` so that this pattern is possible without _forcing_ users to use mixins.

**Comment by nex3** [comment]
Looking into this deeper: a `@layer` statement is actually [allowed anywhere in a stylesheet that the block form is](https://drafts.csswg.org/css-cascade-5/#layer-empty), so while it's _allowed_ to be interspersed with `@import`s it's not _required_ to be. That changes the way we'd handle this situation.

Also, a better alternative for handling this might be to put your `@layer` rule into its own file, at which point you don't need to use mixins at all:

```scss
// _layers.scss
@layer reset, components, utility;
```

```scss
// styles.scss
@use 'layers';
@use 'components';
@use 'pages';
```

Or better yet, `@use` layers from the files that specifically refer to the layers in question, since it'll be guaranteed to be emitted before those files.

**Comment by nex3** [comment]
So we've got an interesting design question here. Currently, the following Sass file:

```scss
@layer foo;
@import 'bar.css';
```

produces:

```css
@import 'bar.css';
@layer foo;
```

This is a clear violation of CSS compatibility, since the input is valid CSS as-is and the output has different semantics. On the other hand:

```scss
a {b: c}
@layer foo;
```

produces the correct output, behavior that should be retained. That raises the question of how to handle a case like the following:

```scss
a {b: c}
@layer foo;
@import 'bar.css';
```

Should this `@layer` rule be moved to the top of the file like the `@import`? Or should it remain in place like the style rule? I can see arguments for both options. I think I lean towards the latter, though—I don't want to encourage users to write useless `@import` rules just to force a `@layer` to be hoisted.

**Comment by rebehayan** [comment]
```sass
// _layers.scss
@layer reset, components, utility;
```
```sass
// styles.scss
@use 'layers';
@use 'components';
@use 'pages';
```

I used this method and solved it.
`@import` was not used because it is not recommended by SCSS.
Thank you for your comment.
Hope you have a happy day.

**Comment by nex3** [comment]
Re-opening because we still need to fix the issues described in https://github.com/sass/sass/issues/3842#issuecomment-2070469822

**Comment by Maximillion195** [comment]
I'm also having the issue in [#3842 (comment)](https://github.com/sass/sass/issues/3842#issuecomment-2070469822) i'm trying to import css from a ui library with the specified layer but I need to declare the layer order prior. When the css is generated, the layer order declaration is placed after the imports.

This
```
@layer defaults, uilib, components;
@import '~@ui/lib/Button/Button.css' layer(uilib);
@import '~@ui/lib/reset.css' layer(defaults);
```

Generates this
```
  !*** css ./node_modules/css-loader/dist/cjs.js!./node_modules/@ui/lib/Button/Button.css' (layer uilib) ***!
[class^=button]{box-sizing:border-box}
  !*** css ./node_modules/css-loader/dist/cjs.js!./node_modules/sass-loader/dist/cjs.js!./src/styles/global.scss ***!
@layer defaults, uilib, components;
```

I tried separating them out in files with `@use` imports but it doesn't make a difference.

```
@use 'layers';
@use 'frameworks';
```

**Comment by hammad-awan** [comment]
> So we've got an interesting design question here. Currently, the following Sass file:
> 
> ```css-scss
> @layer foo;
> @import 'bar.css';
> ```
> 
> produces:
> 
> ```css
> @import 'bar.css';
> @layer foo;
> ```
> 
> This is a clear violation of CSS compatibility, since the input is valid CSS as-is and the output has different semantics. On the other hand:
> 
> ```css-scss
> a {b: c}
> @layer foo;
> ```
> 
> produces the correct output, behavior that should be retained. That raises the question of how to handle a case like the following:
> 
> ```css-scss
> a {b: c}
> @layer foo;
> @import 'bar.css';
> ```
> 
> Should this `@layer` rule be moved to the top of the file like the `@import`? Or should it remain in place like the style rule? I can see arguments for both options. I think I lean towards the latter, though—I don't want to encourage users to write useless `@import` rules just to force a `@layer` to be hoisted.

Hoisting through an @import does not work anyway.

I have moved my layers definition to a layers.scss and imported that as the first import at the top of my styles.scss file and it is still placed after all imports in that file. All @import statements are also hoisted above everything even a @forward which must appear first in a styles.scss file.

Isn't this a significant bug since it doesn't actually allow layering to work correctly?

**Comment by bbroekhuizen** [comment]
Experiencing the same issue. Tried the options which are mentioned here, but still can't get it to work. I want to do some future proofing so i don't have to rethink / refactor everything after 1 or 2 years when Dart Sass 3.0 is the new standard, so this is holding me back significantly.

**Comment by hammad-awan** [comment]
I have worked my way around this. I don't know if my situation is similar or applies to anyone else's but I am posting it so that others may possibly glean some benefit from it.

I had to use multiple third party libraries, one of which I had to forward into my sass entry point styles.scss according to its documentation, so I did the following.

I had a first-library-forward.scss which contained a @use "first-library-core" with (...) variable overrides statment and a @forward "first-library" statement after it. This was part of the procedure I followed as part of the first-library documentation.

```sass
@use "first-library-core" with (...);

@forward "first-library";

// additional styles...

```

Then I created a first-library-theme.scss in in order to wrap this third party library in a layer and placed my layer precedence at the top of the file like so:

```sass
@use sass:meta;

@layer second-library-layer, third-library-layer, first-library-layer;

@layer first-library-layer {
    meta.load-css("first-library-forward");
}
```

Then in my styles.scss I did the following:

```sass
@use sass:meta;

@forward "first-library-theme";

@layer second-library-layer {
    meta.load-css("...path to second-library");
};

@meta.load-css("...path to third-library"); // already wrapped in a layer

// ...additional styles
```

So, in short, I used sass:meta instead of import statements to bypass the hoisting. The layer precedence was the first statement in the generated css file styles.css due to the forward and all of the other styles were "imported" correctly in their wrapped or pre-existing layers in the order they appeared in styles.scss.

**Comment by almeister** [comment]
> Looking into this deeper: a `@layer` statement is actually [allowed anywhere in a stylesheet that the block form is](https://drafts.csswg.org/css-cascade-5/#layer-empty), so while it's _allowed_ to be interspersed with `@import`s it's not _required_ to be. That changes the way we'd handle this situation.
> 
> Also, a better alternative for handling this might be to put your `@layer` rule into its own file, at which point you don't need to use mixins at all:
> 
> // _layers.scss
> @layer reset, components, utility;
>  // styles.scss
> @use 'layers';
> @use 'components';
> @use 'pages';
> Or better yet, `@use` layers from the files that specifically refer to the layers in question, since it'll be guaranteed to be emitted before those files.

I am using the layers in a file approach on a current project which works fine and I don’t think it should be forced by design if it can be avoided.

**Comment by almeister** [comment]
> So we've got an interesting design question here. Currently, the following Sass file:
> 
> @layer foo;
> @import 'bar.css';
> produces:
> 
> @import 'bar.css';
> @layer foo;
> This is a clear violation of CSS compatibility, since the input is valid CSS as-is and the output has different semantics. On the other hand:
> 
> a {b: c}
> @layer foo;
> produces the correct output, behavior that should be retained. That raises the question of how to handle a case like the following:
> 
> a {b: c}
> @layer foo;
> @import 'bar.css';
> Should this `@layer` rule be moved to the top of the file like the `@import`? Or should it remain in place like the style rule? I can see arguments for both options. I think I lean towards the latter, though—I don't want to encourage users to write useless `@import` rules just to force a `@layer` to be hoisted.

I’m also suffering the “can’t @use or @forward scss modules inside a css @layer” problem.  So I’m mostly keen to resolve this 🤷‍♂️

Regarding hoisting of a @layer rule.  Would this need to apply to @layer {…} enclosure rules too?

If it were for @layer order declarations only (eg: layer-1, layer-2;) then I would opt for hoisting it.  Because I cannot imagine a case where hoisting the @layer order would break styling.

If it were for @layer {} enclosures (with nested styling), I would go for hoisting too.  In theory, any @layer {} enclosure should result in correct styling, no matter where it’s moved to inside a file, no?  Of course, preserving the order of multiple @layer {} enclosures would be necessary.

**Comment by sg-code** [comment]
I found a workaround similar to that posted by @hammad-awan using `@use sass:meta;` to allow layered (external/node_modules) imports, without the use of the `@forward` at-rule.

Consider I have 3 scss files in the same directory that I wish to layer (normalize, theme, and globals):

```scss
// main.scss

@use 'layers';
@use 'normalize';
@use 'theme';
@use 'globals';
```

Establish layers at the top of the compiled file as suggested by @nex3.

```scss
// _layers.scss

@layer normalize, theme, globals;
```

Import the external/node_module file using [meta.load-css()](https://sass-lang.com/documentation/at-rules/use/#rules-for-a-pkg-importer) within an `@layer` declaration (in this case a local node_module). The ['pkg:' prefix](https://sass-lang.com/documentation/at-rules/use/#rules-for-a-pkg-importer) is only necessary when referring to a file within node_modules.

```scss
// _normalize.scss

@use "sass:meta";

@layer normalize {
  @include meta.load-css("pkg:modern-normalize/modern-normalize.css")
}
```

Similarly, `@layer` additional files imported with `@use`.

```scss
// _theme.scss

@layer theme {
  body {
    color: aquamarine;
  }
}
```

Although I would hope this behavior gets patched so we can natively declare `@layer` rules at the beginning of a SASS document as defined in CSS specifications, this workaround solved the initial problem along with a transpilation related layering issue I was experiencing in NextJS 15.

**Comment by diesel167** [comment]
> We generally recommend wrapping partials in mixins

I have a file called 'mixins.scss' with a huge amount of mixins. Yesterday i easy imported it by @import and it was very easy and logical. After sass update today i can't do it your recommended way because of error 'you cant declare mixins inside mixins'. (if i wrap in mixin the file with mixins).

**Comment by Mushr0000m** [comment]
Is there any plans to be able to use the "simple" syntax ? like this : 

```scss
// master.scss
@layer base, components, utilities;
@layer base {
    @use "base/reset";
    @use "base/layout";
    @use "base/typography";
}
@layer components {
    @use "components/button";
    @use "components/accordion";
    @use "components/input";
}
@layer utilities {
    @use "utilities/foo";
    @use "utilities/bar";
    @use "utilities/baz";
}
```

**Comment by alexc-mindful** [comment]
I would appreciate this in a few cases too.  It looks like the SASS contributors decided against the @use in @layer approach, encouraging authors to write mixins and include them inside layers instead.  Of course you can use @include meta.load-css() as they suggest to workaround this restriction.

https://github.com/sass/sass/issues/3240#issuecomment-1007004992

While I have been following the @mixin advice in my project, I’m finding cases where a @use inside a @layer would be very helpful.  Most of those cases are related to legacy scss blocks that I have no intention of refactoring.

**Comment by wholewheattoast** [comment]
Is there any official guidance as to how someone *should* or should not use layers with sass, or is this something still up in the air?

I naively would have thought something like the approach mentioned in https://github.com/sass/sass/issues/3842#issuecomment-3012156713 would have been the ideal use case here?

**Comment by nex3** [comment]
We know the current state isn't ideal. We've got lots on our plate, but this is something we intend to come up with a better solution for before too much longer.

**Comment by wholewheattoast** [comment]
@nex3 totally understand about having a lot on your plate.  One related question (please let me know if there is a better place for this): 

I'm looking at updating our scss to use layers, mostly to separate resets, some old bootstrap styles (while we are working to update and replace them), and our in house component styles.

In our top screen.scss have something like:
```
@layer resets, bootstrap, etc;
```

Then each of the various partials in wrapped in a layer like:
```
@layer bootstrap { ... }
```

All this seems to work fine, except if the partial defiles a mixin and the mixin is inside the css layer:

```
@use ...

@layer bootstrap {
  ...
  @mixin form-inline { ... }
}
```
Another partial references this with:
``` @include forms.form-inline```

This throws a build error:

```
Error: Module build failed (from ../../../node_modules/sass-loader/dist/cjs.js):
Undefined mixin.
    ╷
298 │     @include forms.form-inline;
    │     ^^^^^^^^^^^^^^^^^^^^^^^^^^
    ╵
```

However moving the mixin definition out of the css layer (between the initial `@use` and then `@layer` seems to work.

Is this expected behavior?

Thank you,

**Comment by nex3** [comment]
Yes, that's expected. Mixins (and functions and variables) are only visible within the enclosing block (that is, the surrounding `{ }`).

**Comment by stevematney** [comment]
I was never able to get any of the above solutions to work for me with Parcel and two different `@layer` statements. (One from a `node_modules` library and one inline.) I ended up writing a simple PostCSS plugin that seems to do the right thing by hoisting these declarations to the top. Sharing here, just in case it's helpful.

```js
/** @type {import('postcss').Plugin} */
module.exports = {
  postcssPlugin: 'postcss-layer-order-fix',
  prepare() {
    const layerDeclarations = [];

    return {
      AtRule(atRule) {
        if (atRule?.nodes?.length) return;

        layerDeclarations.push(atRule);
      },
      OnceExit(root) {
        if (!layerDeclarations.length) return;

        for (const decl of layerDeclarations) {
          decl.remove();
          root.prepend(decl);
        }
      },
    };
  },
}
```
